### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by savchuksavchuk**
[DEMO LINK](https://savchuksavchuk.github.io/react_todo-app-with-api/)

### CODE REVIEW

File: src/components/Header.tsx [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+          title,
+          completed,
+        } = response;
+
+        const todo = {
+          id,
+          userId,
+          title,
+          completed,
+        };
```
Comment (Bohdan-Maksymiuk) [inline src/components/Header.tsx:69]:
```suggestion
        const newTodo = await addTodo(user.id, newTodo);
```
won't it work like this?

File: src/components/TodoItem.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    setIsRenaming(false);
+    setTitle(newTitle);
+    onRename(todo.id, newTitle);
+  };
+
+  useEffect(() => {
+    renameInput.current?.focus();
+
+    document.addEventListener('keydown', (event) => {
```
Comment (Bohdan-Maksymiuk) [inline src/components/TodoItem.tsx:47]: when do you remove this listener?

**Review by Bohdan-Maksymiuk — CHANGES_REQUESTED** [review]
https://user-images.githubusercontent.com/94289038/207808746-5b68cf34-44ca-4f49-aa7c-28910e1fdd19.mov

bug 🙃

File: src/components/TodoItem.tsx (hunk 2 above)
Comment (savchuksavchuk) [reply src/components/TodoItem.tsx]: fixed

File: src/components/Header.tsx (hunk 1 above)
Comment (savchuksavchuk) [reply src/components/Header.tsx]: Sure it would. I thought it would be nice to pick only values we need.

File: src/components/TodoItem.tsx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          defaultChecked={todo.completed}
+          onClick={() => onComplete(todo)}
+        />
+      </label>
+
+      {!isRenaming
+        ? (
```
Comment (Anastasiia-Svintsova) [inline src/components/TodoItem.tsx:76]:
```suggestion
      {!isRenaming ? (
```

File: src/App.tsx [hunk 4]
```diff
@@ -1,223 +1,213 @@
...
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+  const [errorMessage, setErrorMessage] = useState('');
+  const [filter, setFilter] = useState<FilterValues>(FilterValues.ALL);
+  const [isProcessing, setIsProcessing] = useState<number[]>([]);
   const user = useContext(AuthContext);
-  const newTodoField = useRef<HTMLInputElement>(null);
+  const activeTodosCount = todos.filter(todoItem => !todoItem.completed).length;
+  const completedTodosId = todos.filter(todoItem => todoItem.completed)
+    .map(todo => todo.id);
+  const allCompleted = todos.every(todoItem => todoItem.completed);
```
Comment (Anastasiia-Svintsova) [inline src/App.tsx:30]: You can use useMemo here

File: src/App.tsx [hunk 5]
```diff
@@ -1,223 +1,213 @@
...
+      setIsProcessing([]);
+    }
+  }, []);
+
+  const onComplete = useCallback(async (todo: Todo) => {
+    setIsProcessing(prev => [...prev, todo.id]);
+
+    try {
+      const response
+        = await updateTodo(todo.id, { completed: !todo.completed });
```
Comment (Anastasiia-Svintsova) [inline src/App.tsx:107]:
```suggestion
      const response = await updateTodo(
        todo.id, { completed: !todo.completed }
      );
```

**Review by Anastasiia-Svintsova — CHANGES_REQUESTED** [review]
- The user shouldn't be able to add en empty todo

https://user-images.githubusercontent.com/99403198/208895383-1eaa7bf6-f0c8-4ddf-aba5-f0d7370ac335.mov

**Review by qoa11a — APPROVED** [review]

File: src/App.tsx [hunk 6]
```diff
@@ -1,223 +1,217 @@
...
+  const [isAdding, setIsAdding] = useState(false);
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+  const [errorMessage, setErrorMessage] = useState('');
+  const [filter, setFilter] = useState<FilterValues>(FilterValues.ALL);
+  const [isProcessing, setIsProcessing] = useState<number[]>([]);
   const user = useContext(AuthContext);
-  const newTodoField = useRef<HTMLInputElement>(null);
+  const activeTodosCount = todos.filter(todoItem => !todoItem.completed).length;
+  const completedTodosId = todos.filter(todoItem => todoItem.completed)
+    .map(todo => todo.id);
```
Comment (loralevitska) [inline src/App.tsx:29]:
```suggestion
  const completedTodoIds = todos.filter(todoItem => todoItem.completed)
    .map(todo => todo.id);
```

**Review by loralevitska — APPROVED** [review]

**Pull request closed without merging**